app/jobs/regular/chat_channel_delete.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+module Jobs
+  class ChatChannelDelete < ::Jobs::Base
+    def execute(args = {})
+      chat_channel = ::ChatChannel.with_deleted.find_by(id: args[:chat_channel_id])
+
+      # this should not really happen, but better to do this than throw an error
+      if chat_channel.blank?
+        Rails.logger.warn("Chat channel #{args[:chat_channel_id]} could not be found, aborting delete job.")
+        return
+      end
+
+      DistributedMutex.synchronize(
+        "delete_chat_channel_#{chat_channel.id}",
+        validity: 20.minutes
```

Comment:
I have some concerns around this duration because we saying this job can potentially lock up a Sidekiq worker for a long period of time. Therefore, if I am smart what I can do is to create many chat channels with a ton of chat messages and delete them all at once to lock up all the Sidekiq workers. On a multi-site cluster, this might affect availability. 

A safer design might be to enforce some limitations on the queries that we're running below and queue a new job if we know there are more resources we need to clean up.

Reply:
Hmm this is probably just me not understanding how `DistributedMutex` works correctly. Do we even need it at all? I wouldn't think that the deleting of all the records would take long at all, especially since we don't use foreign keys which usually slow down the delete queries.

I would prefer not to have to introduce more complexity by splitting this into multiple jobs or having to do extra cleanups, we should be able to delete everything in one go. I will time deleting a 10k message channel and report back.

Reply:
@tgxworld I don't think this will be a problem at all, I just deleted 10k messages in milliseconds, IMO we come back to it  if it becomes a problem. I do think maybe I can get rid of DistributedMutex though?

```
  ChatChannel Load (0.4ms)  SELECT "chat_channels".* FROM "chat_channels" WHERE "chat_channels"."id" = 18 LIMIT 1
Deleting webhooks and events for channel 18
  TRANSACTION (0.4ms)  BEGIN
  ChatWebhookEvent Destroy (0.4ms)  DELETE FROM "chat_webhook_events" WHERE "chat_webhook_events"."incoming_chat_webhook_id" IN (SELECT "incoming_chat_webhooks"."id" FROM "incoming_chat_webhooks" WHERE "incoming_chat_webhooks"."chat_channel_id" = 18)
  IncomingChatWebhook Destroy (0.2ms)  DELETE FROM "incoming_chat_webhooks" WHERE "incoming_chat_webhooks"."chat_channel_id" = 18
  TRANSACTION (0.4ms)  COMMIT
Deleting drafts and memberships for channel 18
  ChatDraft Destroy (0.4ms)  DELETE FROM "chat_drafts" WHERE "chat_drafts"."chat_channel_id" = 18
  UserChatChannelMembership Destroy (2.1ms)  DELETE FROM "user_chat_channel_memberships" WHERE "user_chat_channel_memberships"."chat_channel_id" = 18
Deleting chat messages, mentions, revisions, and uploads for channel 18
  TRANSACTION (0.5ms)  BEGIN
  ChatMention Destroy (0.7ms)  DELETE FROM "chat_mentions" WHERE "chat_mentions"."chat_message_id" IN (SELECT "chat_messages"."id" FROM "chat_messages" WHERE "chat_messages"."deleted_at" IS NULL AND "chat_messages"."chat_channel_id" = 18)
  ChatMessageRevision Destroy (0.5ms)  DELETE FROM "chat_message_revisions" WHERE "chat_message_revisions"."chat_message_id" IN (SELECT "chat_messages"."id" FROM "chat_messages" WHERE "chat_messages"."deleted_at" IS NULL AND "chat_messages"."chat_channel_id" = 18)
  ChatUpload Destroy (0.6ms)  DELETE FROM "chat_uploads" WHERE "chat_uploads"."chat_message_id" IN (SELECT "chat_messages"."id" FROM "chat_messages" WHERE "chat_messages"."deleted_at" IS NULL AND "chat_messages"."chat_channel_id" = 18)
  ChatMessage Update All (55.4ms)  UPDATE "chat_messages" SET "deleted_by_id" = NULL, "deleted_at" = '2022-03-17 06:45:01.458358' WHERE "chat_messages"."deleted_at" IS NULL AND "chat_messages"."chat_channel_id" = 18
  TRANSACTION (7.7ms)  COMMIT
  0.034136   0.008146   0.042282 (  0.125489)
```

Reply:
Ah it seems like validity is just used to log a warning message if the lock is held for too long. I think we can just rely on the default here which is 60 seconds.

Reply:
Okay will change to 60s :+1: